elastic/kibana
# [Observability] Update breadcrumbs for observability project based navigation

~⚠️ I'm still putting out some fires with tests, but this is ready to start being reviewed.~

## Summary

A continuation of https://github.com/elastic/kibana/pull/196169 for Observability (please read that PR description first).

Related: https://github.com/elastic/kibana/issues/192050

## Overview of changes

There are essentially three types of breadcrumbs - serverless (which is project style), stateful project style (set through spaces settings), and classic style (the old breadcrumbs we've seen for years). Whilst serverless and stateful project style both use the project based style the navigation trees are slightly different, so the breadcrumbs results are not identical [when they derive the "nav crumbs"](https://github.com/elastic/kibana/blob/9577aa980dd1565fba05e34292fb5c0bba692889/packages/core/chrome/core-chrome-browser-internal/src/project_navigation/breadcrumbs.tsx#L55).

Here "project style" will refer to serverless and stateful project style.

In these changes I've, for the most part, tried to refactor things so Observability solutions route their breadcrumbs through the observability-shared `useBreadcrumbs` hook, this way the logic around project style, adding an Observability crumb in classic etc is consolidated in one place.

[For several solutions `absolute` breadcrumbs are being used](https://github.com/elastic/kibana/blob/9577aa980dd1565fba05e34292fb5c0bba692889/packages/core/chrome/core-chrome-browser-internal/src/project_navigation/breadcrumbs.tsx#L46), and this means we'll roughly have the same breadcrumbs across the 3 experiences (bar Observability being prepended). Teams may want to refine this going forward to pass curated breadcrumbs that take into account the navigation derived "nav crumbs" (again, bearing in mind the trees from serverless and project based chrome do differ), and not use absolute mode. APM is an example of this at the moment. Right now this is an 8.16 bug though, so this aims to make things acceptable, but not necessarily perfect.

### APM

- Project style chrome crumbs have been modelled off the serverless ones. The navigation trees here are the same so this should be fine.

### Infra

- The `infra` `useBreadcrumbs` hook has been removed, it was only being used by logs. Logs now goes via the Observability shared hook using `classicOnly`. 

- Metrics (`useMetricsBreadcrumbs` hook) has been slightly amended to route more of it's logic through the shared hook.

### Logs explorer

- Wasn't setting any nested breadcrumbs at the moment so the logic has been simplified to just set some classic crumbs, and defer the rest to the nav crumbs via the shared hook.

### Synthetics 

- Removed custom logic around prepending Observability, adding link handlers etc in favour of the shared hook.


### Alerts / rules / SLOs etc

- Simple breadcrumb needs so these are mostly setting `classicOnly` and deferring to the nav crumbs in project style.

Several solutions have had their usage of the shared hook updated to pass in the `serverless` plugin. This was missing before, so calls to `serverless.setBreadcrumbs` weren't explicitly happening.

## Testing 

- Add the following to your `kibana.dev.yml`:

```yml
xpack.cloud.id: "ftr_fake_cloud_id:aGVsbG8uY29tOjQ0MyRFUzEyM2FiYyRrYm4xMjNhYmM="
xpack.cloud.base_url: "https://cloud.elastic.co"
```

- For testing stateful project style chrome you'll need to go to Stack Management > Spaces and change the solution view:

![Screenshot 2024-10-21 at 12 44 21](https://github.com/user-attachments/assets/e3d9fe64-f79f-4e31-a5b6-45a06ca4915d)

- Set the above to Classic to test classic breadcrumbs.

- As a reviewer please check your solution against the 3 modes.

## Comparison

Before these changes we'd see something like the following in APM:

![Screenshot 2024-10-11 at 10 56 54](https://github.com/user-attachments/assets/4938b31e-9d4a-429e-abf0-add04d69b62a)

Now we'll see something like this in project style:

![Screenshot 2024-10-21 at 12 48 54](https://github.com/user-attachments/assets/0645a3ae-909e-4a70-a077-d9f83bd1d639)
